### PR TITLE
LiveSync support

### PR DESCRIPTION
**Pull request by Fatme**
Should be merged after this PR https://github.com/telerik/mobile-cli-lib/pull/376

### CODE REVIEW

File: lib/commands/livesync.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+///<reference path="../.d.ts"/>
+"use strict";
+
+export class UsbLivesyncCommand implements ICommand {
```
Comment (ligaz) [inline lib/commands/livesync.ts]: There is no need to prefix this with `Usb`. Both iOS and Android support this over the air.

File: lib/providers/device-app-data-provider.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	}
+}
+
+export class AndroidAppIdentifier extends deviceAppDataBaseLib.DeviceAppDataBase implements Mobile.IDeviceAppData {
+	constructor(_appIdentifier: string) {
+		super(_appIdentifier); 
+	}
+	
+	public get deviceProjectRootPath(): string {
+		return `/data/local/tmp/12590FAA-5EDD-4B12-856D-F52A0A1599F2/${this.appIdentifier}`;
```
Comment (rosen-vladimirov) [inline lib/providers/device-app-data-provider.ts]: is this expected identifier `12590FAA-5EDD-4...`

File: lib/providers/device-app-data-provider.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+export class AndroidCompanionAppIdentifier extends deviceAppDataBaseLib.CompanionDeviceAppDataBase implements Mobile.IDeviceAppData {
+	private static APP_IDENTIFIER = "com.telerik.NativeScript";
+	
+	constructor() {
+		super(AndroidCompanionAppIdentifier.APP_IDENTIFIER); 
+	}
+	
+	public get deviceProjectRootPath(): string {
+		return `/mnt/sdcard/Android/data/${this.appIdentifier}/files/12590FAA-5EDD-4B12-856D-F52A0A1599F2`;
```
Comment (rosen-vladimirov) [inline lib/providers/device-app-data-provider.ts]: is this expected identifier `12590FAA-5EDD-4...`

File: lib/providers/device-app-data-provider.ts (hunk 2 above)
Comment (Fatme) [inline lib/providers/device-app-data-provider.ts]: yes :smile_cat:

File: .settings/launch.json [hunk 4]
```diff
@@ -5,7 +5,7 @@
 	"configurations": [
 		{
 			// Name of configuration; appears in the launch configuration drop down menu.
-			"name": "build",
+			"name": "livesync",
```
Comment (teobugslayer) [inline .settings/launch.json]: Can we .gitignore this file, please?

File: lib/providers/device-app-data-provider.ts (hunk 2 above)
Comment (teobugslayer) [inline lib/providers/device-app-data-provider.ts]: Probably it a good idea to give this path a named constant to explain why it is such.

File: lib/services/usb-livesync-service.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	
+	public restartApplication(deviceAppData: Mobile.IDeviceAppData, localToDevicePaths: Mobile.ILocalToDevicePathData[]): IFuture<void> {
+		return (() => {
+			if(this.$options.companion) {
+				let commands = [ this.liveSyncCommands.SyncFilesCommand() ];			
+				this.livesync(deviceAppData.appIdentifier, deviceAppData.deviceProjectRootPath, commands).wait();
+			} else {
+				// TODO: Introduce this.$mobileHelper.correctDevicePath(path: string) { return helpers.stringReplaceAll(path), '\\', '/') }
+				// Then later on this.$mobileHelper.buildDevicePath should call `correctDevicePath` before returning the output
+				this.device.adb.executeShellCommand(`chmod 0777 ${this.$mobileHelper.buildDevicePath(deviceAppData.deviceProjectRootPath, "app")}`).wait();
```
Comment (teobugslayer) [inline lib/services/usb-livesync-service.ts]: Is the commented code needed?

File: lib/services/usb-livesync-service.ts [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+				
+				let commands: string[] = [];
+				
+				let devicePathRoot = `/data/data/${deviceAppData.appIdentifier}/files`;
+				_.each(localToDevicePaths, localToDevicePath => {
+					let devicePath = helpers.stringReplaceAll(path.join(devicePathRoot, localToDevicePath.getRelativeToProjectBasePath()), '\\', '/');
+					//commands.push(`ls "${path.dirname(devicePath)}" >/dev/null 2>/dev/null || mkdir "${path.dirname(devicePath)}"`);
```
Comment (teobugslayer) [inline lib/services/usb-livesync-service.ts]: Is the commented code needed?

File: .settings/launch.json (hunk 4 above)
Comment (rosen-vladimirov) [inline .settings/launch.json]: :+1: , only basic setup of the file is required, which is already done, all other changes must be gitignored

File: lib/commands/livesync.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+		private $usbLiveSyncService: IUsbLiveSyncService,
+		private $mobileHelper: Mobile.IMobileHelper) { }
+
+	public execute(args: string[]): IFuture<void> {
+		return this.$usbLiveSyncService.liveSync(args[0]);
+	}
+	
+	public canExecute(args: string[]): IFuture<boolean> {
+		return (() => {
+			let platform = args[0];
```
Comment (rosen-vladimirov) [inline lib/commands/livesync.ts]: This way the command will work no matter of the number of arguments passed, for example
`tns livesync android goshko ios peshko john`

File: lib/services/android-debug-service.ts [hunk 8]
```diff
@@ -1,17 +1,41 @@
...
+	
 	constructor(private $devicesServices: Mobile.IDevicesServices,
 		private $platformService: IPlatformService,
 		private $platformsData: IPlatformsData,
 		private $projectData: IProjectData,
 		private $logger: ILogger,
-		private $options: IOptions) { }
+		private $options: IOptions,
+		private $childProcess: IChildProcess,
+        private $mobileHelper: Mobile.IMobileHelper,
```
Comment (rosen-vladimirov) [inline lib/services/android-debug-service.ts]: mixed tabs and spaces

File: lib/services/android-debug-service.ts [hunk 9]
```diff
@@ -27,27 +51,183 @@ class AndroidDebugService implements IDebugService {
...
+	
+	private printDebugPort(packageName: string): IFuture<void> {
+        return (() => {
+            let res = this.$childProcess.spawnFromEvent(this.$staticConfig.getAdbFilePath().wait(), ["shell", "am", "broadcast", "-a", packageName + "-GetDgbPort"], "exit").wait();
+            this.$logger.info(res.stdout);
+        }).future<void>()();
+    }
+	
+	private attachDebugger(packageName: string): void {
+        let startDebuggerCommand = `am broadcast -a \"${packageName}-Debug\" --ez enable true`;
```
Comment (rosen-vladimirov) [inline lib/services/android-debug-service.ts]: do you need to escape the `"`?

File: lib/services/android-debug-service.ts [hunk 10]
```diff
@@ -27,27 +51,183 @@ class AndroidDebugService implements IDebugService {
...
+        } else {
+            let res = this.$childProcess.spawnFromEvent(this.$staticConfig.getAdbFilePath().wait(), ["shell", "am", "broadcast", "-a", packageName + "-Debug", "--ez", "enable", "true"], "exit").wait();
+            let match = res.stdout.match(/result=(\d)+/);
+            if (match) {
+                port = match[0].substring(7);
+            } else {
+                port = 0;
+            }
+        }
+        if ((0 < port) && (port < 65536)) {
```
Comment (rosen-vladimirov) [inline lib/services/android-debug-service.ts]: so we allow ports below 1024?

File: lib/services/usb-livesync-service.ts [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+		$deviceAppDataFactory: Mobile.IDeviceAppDataFactory,
+		$logger: ILogger,
+		private $injector: IInjector,
+		$dispatcher: IFutureDispatcher) {
+			super($devicesServices, $mobileHelper, $localToDevicePathDataFactory, $logger, $options, $deviceAppDataFactory, $fs, $dispatcher); 
+	}
+	
+	public liveSync(platform: string): IFuture<void> {
+		return (() => {
+			this.$options.justlaunch = true;
```
Comment (rosen-vladimirov) [inline lib/services/usb-livesync-service.ts]: Can we set this in the command instead of here.

File: lib/services/android-debug-service.ts (hunk 9 above)
Comment (Fatme) [inline lib/services/android-debug-service.ts]: I guess yes :smile_cat:  @slavchev what do you think?

File: lib/services/android-debug-service.ts (hunk 10 above)
Comment (Fatme) [inline lib/services/android-debug-service.ts]: @slavchev what do you think?

**Comment by rosen-vladimirov** [comment]
:+1:

**Pull request merged**